### PR TITLE
FSDP: Fix saving and loading checkpoints with use_sharded_state=True

### PR DESCRIPTION
**Pull request by shruti-bh**
* fairseq's FSDP wrapper's state_dict() calls fairscale's FSDP module's local_state_dict() 
and fairscale's FSDP module's local_state_dict() calls self.state_dict() which calls fairseq's FSDP wrapper's state_dict() , thus causing infinite recursion.
* using super().state_dict() did not work for me - it continued calling the fairseq FSDP wrapper's state_dict()
* using super(FullyShardedDataParallel, self).state_dict() called nn.Module's state_dict() method
* I finally gave up and did the following - 
   * created _state_dict() and _load_state_dict() so that I have a way to call fairscale's FSDP module's state_dict() and load_state_dict() without any ambiguity

This works correctly for a dense multilingual MT model in fairseq
* continuing training after saving checkpoint
```
2021-04-02 16:41:29 | INFO | train_inner | {"epoch": 1, "update": 0.011, "loss": "14.259", "nll_loss": "13.983", "ppl": "16197.5", "wps": "10226.9", "ups": "0.26", "wpb": "40069", "bsz": "1896", "num_updates": "81", "lr": "2.0348e-05", "gnorm": "0.915", "loss_scale": "16", "train_wall": "1", "gb_free": "10.4", "wall": "108"}
2021-04-02 16:41:29 | INFO | train_inner | {"epoch": 1, "update": 0.011, "loss": "14.194", "nll_loss": "13.911", "ppl": "15406.4", "wps": "78191.6", "ups": "1.81", "wpb": "43224", "bsz": "2224", "num_updates": "82", "lr": "2.05979e-05", "gnorm": "0.96", "loss_scale": "16", "train_wall": "0", "gb_free": "10.6", "wall": "108"}
```
* reloading checkpoint and continuing training
```
2021-04-02 16:42:58 | INFO | train_inner | {"epoch": 1, "update": 0.011, "loss": "14.259", "nll_loss": "13.983", "ppl": "16197.5", "wps": "1768", "ups": "0.04", "wpb": "40069", "bsz": "1896", "num_updates": "81", "lr": "2.0348e-05", "gnorm": "0.915", "loss_scale": "16", "train_wall": "8", "gb_free": "10.6", "wall": "0"}
2021-04-02 16:42:58 | INFO | train_inner | {"epoch": 1, "update": 0.011, "loss": "14.194", "nll_loss": "13.911", "ppl": "15406.4", "wps": "57383.1", "ups": "1.33", "wpb": "43224", "bsz": "2224", "num_updates": "82", "lr": "2.05979e-05", "gnorm": "0.96", "loss_scale": "16", "train_wall": "1", "gb_free": "10.7", "wall": "0"}
```

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:552]: could you please add a comment on why is this wrapper needed? without it, it is easily for future refactor to undo this change.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:566]: I think you don't need to introduce the `_state_dict` function if you do the following:

`FSDP.state_dict(self, *args, *kwargs)`

This ensures that you will call the right version state_dict from this function. It should be the same as the changes you have done here and should be simpler.

I don't know if this is consider pythonic though. Seems a like simpler and clearer method to me.

Comment (shruti-bh) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:566]: Since `state_dict()` is an object level method and not a class method, `FSDP.state_dict()` won't work in this case.

I had hoped `super().state_dict()` calls the fairscale version. But in practice, it doesn't call the fairscale version as explained in the PR description.
I will try to dig into why that might be the case.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (shruti-bh) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:552]: Sure I will add a comment - good point

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:566]: Have you tried `FSDP.state_dict()`? It should work since you give the `self` argument explicitly. I think doing that would explicitly call the right version of it.

**Review by sshleifer — APPROVED** [review]
LGTM, pending lint, Min's asks.
```
fs_test () {
	black . && isort . && flake8 && mypy --ignore-missing-imports --scripts-are-modules --pretty . && echo ALL GOOD
}
```

**Comment by sshleifer** [comment]
Bonus points for some sort of regression test where you make an `FSDPWrapper` class in `pytest tests/nn/data_parallel/` and show the bug fix!

**Comment by min-xu-ai** [comment]
@shruti-bh, do you want me to take over this PR and merge it?

**Comment by shruti-bh** [comment]
> @shruti-bh, do you want me to take over this PR and merge it?

Hi @min-xu-ai That would be awesome if you can take this over! Sorry I have been swamped with some other stuff.

**Comment by min-xu-ai** [comment]
> > @shruti-bh, do you want me to take over this PR and merge it?
> 
> Hi @min-xu-ai That would be awesome if you can take this over! Sorry I have been swamped with some other stuff.

No worries. I may not get to it this week, but will do.

**Comment by min-xu-ai** [comment]
sorry for the delay, don't merge this. I will get to this soon.

**Comment by min-xu-ai** [comment]
@shruti-bh new version is ready. :-)

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:694]: the key fix is this line.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:994]: these cuda guards are added to make cpu-only test happy

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:694]: 🔥

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:678]: Are you doing this all to please mypy? We need both overloads, I assume?

**Review by sshleifer — APPROVED** [review]

File: tests/nn/data_parallel/test_fsdp_state_dict.py
Comment (sshleifer) [inline tests/nn/data_parallel/test_fsdp_state_dict.py:247]: why skip_if_cuda?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:678]: yeah, it is *the official way* to do overloaded function as far as I can tell. if we only have the (any, any)->any signature, it doesn't match the parent class signature and I suppose this will help typing checking our code a tiny bit better.

File: tests/nn/data_parallel/test_fsdp_state_dict.py
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_state_dict.py:247]: our cuda test is pretty long already. I hope to save a bit time since this test is covering a code path that's common on cpu/gpu anyway.

**Pull request merged**
